File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    TreeNode prev;
    public boolean isValidBST(TreeNode root) {
        // do an inorder traversal, if the result is in sorted order it means that the tree is BST. 
        return inorder(root);
    }

    private boolean inorder(TreeNode root) {
        if (root != null) {
            if (!inorder(root.left)) {
                return false;
            }
            if (prev != null && root.val <= prev.val) { // No need of using extra space as we can get out of inorder traversal if the current value is less than the previous node.
                return false;
            }
            prev = root;

            return inorder(root.right);
        }
        return true;
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        TreeNode* prev = nullptr; // pointer to keep track of the previous node in in-order traversal
        return inorder(root, prev);
    }

private:
    bool inorder(TreeNode* node, TreeNode*& prev) {
        if (!node) return true;

        if (!inorder(node->left, prev)) return false;

        if (prev && node->val <= prev->val) return false;

        prev = node;

        return inorder(node->right, prev);
    }
};